File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/toml/TomlDocumentMapper.java
package de.lino.database.database.nosql.toml;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.moandjiezana.toml.Toml;
import com.moandjiezana.toml.TomlWriter;
import de.lino.database.json.JsonDocument;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Translates between this driver's {@link JsonDocument} model and TOML text, for the TOML file
 * store. The two models do not overlap perfectly, and this mapper resolves every mismatch in
 * one place - so {@link TOMLDatabaseSection}'s storage primitives stay plain file operations -
 * with the following deliberate rules:
 * <ul>
 *   <li><b>{@code null} values are dropped on write.</b> TOML has no {@code null} literal at
 *       all; silently omitting the key (rather than failing the whole write) mirrors how
 *       {@link JsonDocument#append(String, String)} already treats a {@code null} value as
 *       "nothing to store".</li>
 *   <li><b>Integral JSON numbers stay integral.</b> A naive Gson round trip would read every
 *       number back as {@code double} and rewrite {@code age = 23} as {@code age = 23.0};
 *       numbers are therefore converted by inspecting their exact value - integral and
 *       long-ranged becomes a TOML integer, everything else a TOML float.</li>
 *   <li><b>TOML's own restrictions apply to what documents can be stored.</b> Arrays must be
 *       homogeneous and special floating-point values ({@code NaN}, infinities) do not exist
 *       in TOML 0.4; a document violating them fails at write time rather than producing a
 *       file that can never be parsed back.</li>
 * </ul>
 */
final class TomlDocumentMapper {

    private TomlDocumentMapper() {
    }

    /**
     * Renders {@code document} as TOML text, applying the conversion rules above.
     *
     * @param document the document to render
     * @return the TOML representation
     */
    static @NotNull String toToml(@NotNull JsonDocument document) {
        return new TomlWriter().write(toTomlValue(document.getJsonObject()));
    }

    /**
     * Parses TOML text back into a {@link JsonDocument}, the inverse of {@link #toToml}. TOML
     * types map naturally onto JSON ones (tables to objects, arrays to arrays, integers and
     * floats to numbers); the one one-way street is a TOML datetime, which JSON lacks and
     * which comes back as its serialized string form.
     *
     * @param toml the TOML text to parse
     * @return the parsed document
     * @throws IllegalStateException if {@code toml} is not valid TOML (toml4j's parse failure)
     */
    static @NotNull JsonDocument fromToml(@NotNull String toml) {
        return new JsonDocument(new JsonDocument().getGson().toJson(new Toml().read(toml).toMap()));
    }

    /**
     * Recursively converts a Gson tree into the plain-Java shape ({@link Map}/{@link List}/
     * boxed primitives) toml4j's writer expects, enforcing the class rules: {@code null}s are
     * dropped, integral numbers stay {@code long}.
     *
     * @param element the element to convert
     * @return the converted value, or {@code null} for JSON {@code null} (dropped by the
     * object/array branches)
     */
    private static @Nullable Object toTomlValue(@NotNull JsonElement element) {

        if (element.isJsonObject()) {

            final JsonObject object = element.getAsJsonObject();
            final Map<String, Object> map = new LinkedHashMap<>();

            for (final Map.Entry<String, JsonElement> entry : object.entrySet()) {
                final Object value = toTomlValue(entry.getValue());
                if (value != null) map.put(entry.getKey(), value);
            }

            return map;

        }

        if (element.isJsonArray()) {

            final JsonArray array = element.getAsJsonArray();
            final List<Object> list = new ArrayList<>(array.size());

            for (final JsonElement item : array) {
                final Object value = toTomlValue(item);
                if (value != null) list.add(value);
            }

            return list;

        }

        if (element.isJsonPrimitive()) {

            final JsonPrimitive primitive = element.getAsJsonPrimitive();

            if (primitive.isBoolean()) return primitive.getAsBoolean();
            if (primitive.isString()) return primitive.getAsString();

            return toTomlNumber(primitive);

        }

        return null; // JsonNull - dropped by the caller, TOML cannot express it

    }

    /**
     * Converts a JSON number to the boxed type that keeps its TOML rendering faithful:
     * anything with an exact integral value inside {@code long} range becomes a TOML integer,
     * everything else a TOML float - see the class documentation for why a plain
     * double-everything conversion is not acceptable.
     *
     * @param primitive the numeric primitive to convert
     * @return the boxed {@link Long} or {@link Double}
     */
    private static @NotNull Object toTomlNumber(@NotNull JsonPrimitive primitive) {

        try {

            final BigDecimal exact = primitive.getAsBigDecimal().stripTrailingZeros();
            if (exact.scale() <= 0) return exact.longValueExact();
            return exact.doubleValue();

        } catch (final ArithmeticException | NumberFormatException outOfIntegerReach) {
            // Beyond long range, or a special value like NaN: hand the double form to the
            // writer - TOML 0.4 cannot express NaN/Infinity, so those fail there, loudly.
            return primitive.getAsDouble();
        }

    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/toml/toml_document_mapper.py
"""Mirror of ``de.lino.database.database.nosql.toml.TomlDocumentMapper``."""

from __future__ import annotations

import math
import tomllib
from typing import Any

import tomli_w
from database_driver.api.json.json_document import JsonDocument


def to_toml(document: JsonDocument) -> str:
    """Renders ``document`` as TOML text.

    The JSON and TOML models do not overlap perfectly, and this mapper resolves every
    mismatch in one place - so the TOML section's storage primitives stay plain file
    operations - with the following deliberate rules, identical to the Java edition's:

    - **``None`` values are dropped on write.** TOML has no ``null`` literal at all;
      silently omitting the key (rather than failing the whole write) mirrors how
      ``JsonDocument.append`` already treats a ``None`` value as "nothing to store".
    - **Integral JSON numbers stay integral.** The Java edition needs explicit handling
      because Gson reads every number back as ``double``; Python's ``json`` module keeps
      ``int`` and ``float`` distinct natively, so ``age = 23`` round-trips as a TOML
      integer without extra work - the rule holds by construction.
    - **TOML's own restrictions apply to what documents can be stored.** Special
      floating-point values (``NaN``, infinities) are rejected here at write time rather
      than producing a file that can never be parsed back, and heterogeneous arrays are
      constrained by the writer itself.

    Raises:
        ValueError: If the document holds a value TOML cannot express.
    """
    return tomli_w.dumps(_to_toml_value(document.json_object))


def from_toml(toml: str) -> JsonDocument:
    """Parses TOML text back into a ``JsonDocument``, the inverse of :func:`to_toml`.
    TOML types map naturally onto JSON ones (tables to objects, arrays to arrays,
    integers and floats to numbers); the one one-way street is a TOML datetime, which
    JSON lacks and which comes back as its serialized string form.

    Raises:
        tomllib.TOMLDecodeError: If ``toml`` is not valid TOML.
    """
    parsed = tomllib.loads(toml)
    return JsonDocument(_from_toml_value(parsed))


def _to_toml_value(value: Any) -> Any:
    """Recursively converts a JSON-native tree into the shape the TOML writer expects,
    enforcing the mapper rules: ``None``s are dropped, non-finite floats are rejected."""
    if isinstance(value, dict):
        converted = {}
        for key, entry in value.items():
            entry_value = _to_toml_value(entry)
            if entry_value is not None:
                converted[key] = entry_value
        return converted
    if isinstance(value, list):
        return [entry for entry in (_to_toml_value(item) for item in value) if entry is not None]
    if isinstance(value, float) and not math.isfinite(value):
        raise ValueError(f"@TomlDocumentMapper: TOML cannot express {value}")
    return value


def _from_toml_value(value: Any) -> Any:
    """Recursively converts a parsed TOML tree into JSON-native values; datetimes and
    dates - which JSON lacks - come back as their serialized string form, matching the
    Java edition's round-trip behavior."""
    if isinstance(value, dict):
        return {key: _from_toml_value(entry) for key, entry in value.items()}
    if isinstance(value, list):
        return [_from_toml_value(entry) for entry in value]
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    return str(value)  # TOML datetime/date/time: JSON has no counterpart
